rust/libnewsboat-ffi/src/lib.rs

```diff
@@ -58,6 +58,19 @@ pub extern "C" fn rs_to_u(in_str: *const c_char, default_value: u32) -> u32
     utils::to_u( rs_str, default_value)
 }
 
+#[no_mangle]
+pub extern "C" fn rs_absolute_url(base_url: *const c_char, link: *const c_char) -> *mut c_char {
+    let rs_base_url = unsafe { CStr::from_ptr(base_url) };
+    let rs_base_url = rs_base_url.to_string_lossy();
+    let rs_link = unsafe { CStr::from_ptr(link) };
+    let rs_link = rs_link.to_string_lossy();
+
+    match utils::absolute_url(&rs_base_url, &rs_link) {
+        Err(_) => std::ptr::null_mut(),
```

Comment:
> C++ implementation returns link if xmlBuildURI returned an error, which seems a valid choice to me considering the usage of this function: it's called during RSS/Atom parsing and turns relative URLs into absolute ones. If the base (base_url) is not specified, it makes sense to return the original (relative) link rather than nullptr. I'd replicate that in Rust.

Ah that makes sense. Changed it to return link in case of any error. Still, I wonder how we could map `Result<T, E>` to C++ in general.

> I hope to finish moving to Rust long before it becomes viable to use C++20 in Newsboat (or whichever Standard gets std::expected) :)

Yeah, I was half joking with that one :wink:

Reply:
> Still, I wonder how we could map Result<T, E> to C++ in general.

I'd use [`expected` header-only library](https://github.com/TartanLlama/expected), and fling an `(is_err: bool, contents: *const c_void)` over FFI. Both C++ and Rust will have to do casting/transmuting to convert that pointer from/to an actual value.

Now that I think about it more, your question don't seem as hypothetical anymore. Some parts of Newsboat are pretty heavy on exceptions, and sooner or later we'll have to convert them to return codes (or something like thta) in order to port them to Rust. Perhaps porting to `expected` would be a better way to do it.